src/wrapper/jnienv.rs

```diff
@@ -347,6 +347,38 @@ impl<'a> JNIEnv<'a> {
         Ok(global)
     }
 
+    /// Turns an object into a weak global ref.
+    ///
+    /// `WeakGlobalRef` doesn't prevent the underlying object from been garbage
+    /// collected. It's not safe to access the underlying object via `WeakGlobalRef`.
+    /// Instead, consider using [upgrade_weak_global_ref](#method.upgrade_weak_global_ref) for
+    /// create `GlobalRef` and check the `GlobalRef` is not null.
+    pub fn new_weak_global_ref<O>(&self, obj: O) -> Result<WeakGlobalRef>
+    where
+        O: Into<JObject<'a>>,
+    {
+        let new_ref: JWeak =
+            jni_unchecked!(self.internal, NewWeakGlobalRef, obj.into().into_inner()).into();
+        let weak_global =
+            unsafe { WeakGlobalRef::from_raw(self.get_java_vm()?, new_ref.into_inner()) };
+        Ok(weak_global)
+    }
+
+    /// Try to upgrade a weak global ref into a global ref.
+    ///
+    /// Returned `GlobalRef` may be null if the underlying object was garbage collected.
+    pub fn upgrade_weak_global_ref(&self, weak_global_ref: &WeakGlobalRef) -> Result<GlobalRef> {
```

Comment:
What do you think about returning `Result<Option<GlobalRef>>` instead of `Result<GlobalRef>`? This way `GlobalRef` will always be non-null and the api becomes a little more obvious.

Reply:
I agree we should avoid null-checks. From the other side Result<Option<GlobalRef>> looks a little bit too verbose. 

Do you think it's possible to use one of existing [ErrorKind](https://docs.rs/jni/0.17.0/jni/errors/enum.ErrorKind.html)s to provide meaningful error as a part of `Result<GlobalRef>`. It'll allow us to keep the method signature `pub fn upgrade_weak_global_ref(&self, weak_global_ref: &WeakGlobalRef) -> Result<GlobalRef>`. Or we can create a new one to cover that case.

What do you think?

Reply:
I can't comment on the ergonomics in Rust, but Option<GlobalRef> may relieve the users from comparing the returned reference to `null`. The thing is, `null` GlobalRef/AutoLocal is usually a possible thing, but WeakGlobalRef are not really supposed to be `null`. We may consider distinguishing between "WGR was _originally_ null" (but maybe, we shall rather make such case _non-representable_ by returning an Option<WGR> from `new_weak_global_ref`?) and "WGR wasn't null, but the object was GCed". Alternatively, we may keep things as in JNI, where WGR could be `null`, and you can get `null` in `NewLocalRef`/`NewGlobalRef` which could mean two things (WGR originally null or the object got GCed).